src/Kaleidoscope-LEDEffect-Rainbow.h

```diff
@@ -19,26 +19,30 @@ class LEDRainbowEffect : public LEDMode {
   uint16_t rainbow_ticks = 10;  //  delays between update
 
   byte rainbow_saturation = 255;
-  byte rainbow_value = 50;
+  byte rainbow_value = 151;
 };
 
 
 class LEDRainbowWaveEffect : public LEDMode {
  public:
-  LEDRainbowWaveEffect(void) {}
+  LEDRainbowWaveEffect(void) {} //  Empty constructor with default values
+  LEDRainbowWaveEffect(byte);
+  LEDRainbowWaveEffect(byte, uint16_t);
+  LEDRainbowWaveEffect(byte, uint16_t, uint8_t);
```

Comment:
Are the extra constructors useful? Most of the time, one will just use the global `LEDRainbowEffect` object, which uses the empty constructor. I see little reason for having two rainbow effects in the same sketch...

So `LEDRainbowEffect.delay(N)` and/or `LEDRainbowEffect.precision(X)` from the `setup()` method of the sketch would be enough to tweak the effect.

What do you think?

Reply:
I actually have 8 of them in my sketch for my board. The default one moves too slowly for my preference, so I have two sets that move medium-fast (only changing default delay) and super-fast (changing both delay and precision in the constructor). Then I have one each of those sets at 25%, 50%, 75%, and 100% brightness. Then I can cycle through those with the next/previous LED effect key.

We don't have a system-defined way of universally increasing or decreasing LED brightness, or a system of hooks for globally tweaking current backlight mode changes (say, an API with defined global keys that plugins can hook into that will always work on the current LED backlighting mode).

I'm fine with having simply the void constructor and then the 3-parameter constructor, but in general my design philosophy as an engineer is that you should provide as many overloaded constructors or methods with sane default values as you think your end-user may need. It's better to be inclusive  rather than exclusive, in general. But I don't typically have to worry about size or memory constraints so much, either, so I am willing to concede on that point here. :-)

Reply:
To further add to that, I think the order of the constructors is also relevant. I think the most common use case will be having multiple objects with different backlight levels (first constructor with 1 parameters for backlight). After that, I think people will want to change the backlight level and speed. Finally, I think the last parameters that people will want to change is the precision, to fine-tune their settings.

To me, I can definitely foresee somebody wanting each of those constructors, as I actually have used them all as I was tweaking settings to my liking.

Reply:
Ah, I see. I didn't think of your use case... well, I did, but my solution for the same goal would have been a bit different, and perhaps less straightforward: I'd have used a different key for brightness manipulation, a macro that would check the active mode, and cycle through pre-defined brigthness values (for modes that support it).

In the long run, I do agree: a way to globally set brightness would be useful.

In the short term, I'd think that having one constructor would be more in-line with how we currently do things, and brigthness would be set in a mode-specific way, if so desired. That saves quite a lot of resources (RAM and PROGMEM alike), at the cost of a bit more code to write. This would also be easier to migrate from once we have global brightness control, I think...

@obra, what do you think?

Reply:
I was actually just about to implement global brightness in LEDControl. I am going to implement an rgbToHsv() function, and override setLed to use the given rgb value and a global max brightness to find the equivalent rgb hue/sat at the given global brightness and use that instead. Then provide some global hooks for board brightness and keys to lower, increase, and disable backlight.